File: justViewer/project_JustViewer/src/main/java/com/kds3393/just/viewer/provider/DBProvider.java
package com.kds3393.just.viewer.provider;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import android.database.sqlite.SQLiteOpenHelper;
import android.database.sqlite.SQLiteStatement;
import android.util.Log;

import com.common.utils.debug.CLog;

/**
 * Created by android on 2016-09-13.
 */
public class DBProvider {
    private static final String TAG = "DBProvider";

    private SQLiteDatabase m_db;
    private static DBHelper m_dbHelper;

    public DBProvider(Context context, SQLiteDatabase.CursorFactory factory, int version) {
        m_dbHelper = new DBHelper(context, context.getDatabasePath(DBInfo.DB_NAME).getPath(), null, DBInfo.DATABASE_VERSION);
        m_db = m_dbHelper.getWritableDatabase();
    }

    public class DBHelper extends SQLiteOpenHelper {
        DBHelper(Context context, String name, SQLiteDatabase.CursorFactory factory, int version) {
            super(context, name, factory, version);
        }

        @Override
        public void onCreate(SQLiteDatabase db) {
            db.beginTransaction();
            try {
                db.execSQL(DBInfo.IMAGE_FAVORITE_TABLE_CREATE);
                db.setTransactionSuccessful();
            } catch (SQLiteException e) {
                CLog.e(TAG, e);
            } finally {
                db.endTransaction();
            }

            db.beginTransaction();
            try {
                db.execSQL(DBInfo.BOOKMARK_TABLE_CREATE);
                db.setTransactionSuccessful();
            } catch (SQLiteException e) {
                CLog.e(TAG, e);
            } finally {
                db.endTransaction();
            }
        }

        @Override
        public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        }
    }

    public DBHelper getDBHelper() {
        return m_dbHelper;
    }

    public boolean isOpened() {
        if (m_db == null) {
            return false;
        }
        return m_db.isOpen();
    }

    public void close() {
        if (m_db != null) {
            m_db.close();
        }
    }

    /**
     * @brief DB에 data insert
     * @param strTableName
     * @param values
     * @return the row ID of the newly inserted row, or -1 if an error occurred
     */
    public long insertDB(String strTableName, ContentValues values) {
        if (isOpened()) {
            return m_db.insert(strTableName, null, values);
        }
        return -1;
    }

    /**
     * @brief DB에 data update
     * @param strTableName
     * @param values
     * @param strWhere
     * @return
     */
    public int updateDB(String strTableName, ContentValues values, String strWhere) {
        if (isOpened()) {
            Log.i(TAG, "DB UPDATE : " + "strTableName= " + strTableName + values.toString() + " where=" + strWhere);
            return m_db.update(strTableName, values, strWhere, null);
        }
        return -1;
    }

    /**
     * @brief DB data delete
     * @param strTableName
     * @param strWhere
     * @return the number of rows affected if a whereClause is passed in, 0 otherwise.
     *         To remove all rows and get a count pass "1" as the whereClause.
     */
    public long deleteDB(String strTableName, String strWhere) {
        if (strWhere == null) {
            strWhere = "";
        }
        if (isOpened()) {
            Log.i(TAG, "DB DELETE : " + "strTableName= " + strTableName + " where=" + strWhere);
            return m_db.delete(strTableName, strWhere, null);
        }
        return -1;
    }

    /**
     * @brief DB에서 원하는 Field를 query 한다.
     * @param strTable
     * @param strSelect
     * @param strWhere
     * @return
     */
    public Cursor getDBData(String strTable, String[] strSelect, String strWhere) {
        if (isOpened()) {
            Cursor cursor = m_db.query(strTable, strSelect, strWhere, null, null, null, null);
            return cursor;
        }
        return null;
    }

    /**
     * @brief DB에서 원하는 Field를 정렬 query 한다.
     * @param strTable
     * @param strSelect
     * @param strWhere
     * @return
     */
    public Cursor getDBData(String strTable, String[] strSelect, String strWhere, String orderby) {
        if (isOpened()) {
            Cursor cursor = m_db.query(strTable, strSelect, strWhere, null, null, null, orderby);
            return cursor;
        }

        return null;
    }

    /**
     * @brief rawQuery 실행
     * @param strQuery
     * @return
     */
    public Cursor rawQuery(String strQuery) {
        if (isOpened()) {
            Cursor cursor = m_db.rawQuery(strQuery, null);
            return cursor;
        }
        return null;
    }

    /**
     * @brief execSQL 실행
     * @param strQuery
     * @return
     */
    public void execSQL(String strQuery) {
        if (isOpened()) {
            m_db.execSQL(strQuery);
        }
    }

    /**
     * @brief 테이블 존재여부 검사
     * @param strTableName
     *            검사할 테이블 명
     * @return 존재여부 (true : 존재, false : 존재하지 않음)
     */
    public boolean isExistTable(String strTableName) {
        if (isOpened()) {
            boolean bRet = false;
            String temp = "SELECT name FROM sqlite_master WHERE name='" + strTableName + "'";
            Cursor cursor = m_db.rawQuery(temp, null);

            if (cursor != null) {
                if (cursor.getCount() > 0) {
                    bRet = true;
                }
                cursor.close();
            }

            return bRet;
        }
        return false;
    }

    public boolean isExistData(String strTableName, String idFieldName, int id) {
        if (isOpened()) {
            boolean bRet = false;
            String temp = "SELECT * FROM " + strTableName + " WHERE " + idFieldName + "='" + id + "'";
            Cursor cursor = m_db.rawQuery(temp, null);

            if (cursor != null) {
                if (cursor.getCount() > 0) {
                    bRet = true;
                }
                cursor.close();
            }

            return bRet;
        }
        return false;
    }

    public boolean isExistData(String strTableName, String idFieldName, int id, String userfieldName, String userid) {
        if (isOpened()) {
            boolean bRet = false;
            String temp = "SELECT * FROM " + strTableName + " WHERE " + idFieldName + "='" + id + "' AND " + userfieldName + "='" + userid + "'";
            Cursor cursor = m_db.rawQuery(temp, null);

            if (cursor != null) {
                if (cursor.getCount() > 0) {
                    bRet = true;
                }
                cursor.close();
            }

            return bRet;
        }
        return false;
    }

    public void startTransaction() {
        if (isOpened()) {
            m_db.beginTransaction();
        }
    }

    public void transactionSuccess() {
        if (isOpened()) {
            m_db.setTransactionSuccessful();
        }
    }

    public void endTransaction() {
        if (isOpened()) {
            m_db.endTransaction();
        }
    }

    public SQLiteStatement compileStatement(String sql) {
        return m_db.compileStatement(sql);
    }
}

File: JustViewer2/app/src/main/java/com/kds3393/just/justviewer2/db/DBProvider.kt
package com.kds3393.just.justviewer2.db

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteDatabase.CursorFactory
import android.database.sqlite.SQLiteException
import android.database.sqlite.SQLiteOpenHelper
import android.database.sqlite.SQLiteStatement
import common.lib.debug.CLog

class DBProvider(context: Context?, path: String?, factory: CursorFactory?, version: Int) {
    val db: SQLiteDatabase
    var dbHelper: DBHelper
    init {
        dbHelper = DBHelper(context, path + "/" + DBInfo.DB_NAME, null, DBInfo.DATABASE_VERSION)
        db = dbHelper.writableDatabase
    }

    inner class DBHelper internal constructor(context: Context?, name: String?, factory: CursorFactory?, version: Int) : SQLiteOpenHelper(context, name, factory, version) {
        override fun onCreate(db: SQLiteDatabase) {
            CLog.e("KDS3393_TEST_DB_onCreate")
            db.beginTransaction()
            try {
                db.execSQL(DBInfo.TEXT_BOOKMARK_TABLE_CREATE)
                db.execSQL(DBInfo.FAVO_BOOKMARK_TABLE_CREATE)
                db.execSQL(DBInfo.IMAGE_BOOKMARK_TABLE_CREATE)
                db.setTransactionSuccessful()
            } catch (e: SQLiteException) {
                CLog.e(e)
            } finally {
                db.endTransaction()
            }
        }

        override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
            db.execSQL("DROP TABLE IF EXISTS " + DBInfo.FAVO_BOOKMARK_TABLE)
            db.execSQL("DROP TABLE IF EXISTS " + DBInfo.IMAGE_BOOKMARK_TABLE)
            db.execSQL("DROP TABLE IF EXISTS " + DBInfo.TEXT_BOOKMARK_TABLE)
            onCreate(db)
        }
    }

    val isOpened: Boolean
        get() = if (db == null) {
            false
        } else db.isOpen

    fun close() {
        if (db != null) {
            db.close()
        }
    }

    /**
     * @brief DB에 data insert
     * @param strTableName
     * @param values
     * @return the row ID of the newly inserted row, or -1 if an error occurred
     */
    fun insertDB(strTableName: String, values: ContentValues?): Long {
        return if (isOpened) {
            db.insert(strTableName, null, values)
        } else -1
    }

    /**
     * @brief DB에 data update
     * @param strTableName
     * @param values
     * @param strWhere
     * @return
     */
    fun updateDB(strTableName: String, values: ContentValues, strWhere: String): Int {
        if (isOpened) {
            CLog.i("DB UPDATE : strTableName = $strTableName$values where = $strWhere")
            return db.update(strTableName, values, strWhere, null)
        }
        return -1
    }

    /**
     * @brief DB data delete
     * @param strTableName
     * @param strWhere
     * @return the number of rows affected if a whereClause is passed in, 0 otherwise.
     * To remove all rows and get a count pass "1" as the whereClause.
     */
    fun deleteDB(strTableName: String, strWhere: String?): Long {
        var strWhere = strWhere
        if (strWhere == null) {
            strWhere = ""
        }
        if (isOpened) {
            CLog.i("DB DELETE : strTableName= $strTableName where=$strWhere")
            return db.delete(strTableName, strWhere, null).toLong()
        }
        return -1
    }

    /**
     * @brief DB에서 원하는 Field를 query 한다.
     * @param strTable
     * @param strSelect
     * @param strWhere
     * @return
     */
    fun getDBData(strTable: String, strSelect: Array<String?>?, strWhere: String?): Cursor? {
        return if (isOpened) {
            db.query(strTable, strSelect, strWhere, null, null, null, null)
        } else null
    }

    /**
     * @brief DB에서 원하는 Field를 정렬 query 한다.
     * @param strTable
     * @param strSelect
     * @param strWhere
     * @return
     */
    fun getDBData(strTable: String, strSelect: Array<String?>?, strWhere: String?, orderby: String?): Cursor? {
        return if (isOpened) {
            db.query(strTable, strSelect, strWhere, null, null, null, orderby)
        } else null
    }

    /**
     * @brief rawQuery 실행
     * @param strQuery
     * @return
     */
    fun rawQuery(strQuery: String): Cursor? {
        return if (isOpened) {
            db.rawQuery(strQuery, null)
        } else null
    }

    /**
     * @brief execSQL 실행
     * @param strQuery
     * @return
     */
    fun execSQL(strQuery: String?) {
        if (isOpened) {
            db.execSQL(strQuery)
        }
    }

    /**
     * @brief 테이블 존재여부 검사
     * @param strTableName
     * 검사할 테이블 명
     * @return 존재여부 (true : 존재, false : 존재하지 않음)
     */
    fun isExistTable(strTableName: String): Boolean {
        if (isOpened) {
            var bRet = false
            val temp = "SELECT name FROM sqlite_master WHERE name='$strTableName'"
            val cursor = db.rawQuery(temp, null)
            if (cursor != null) {
                if (cursor.count > 0) {
                    bRet = true
                }
                cursor.close()
            }
            return bRet
        }
        return false
    }

    fun isExistData(strTableName: String, idFieldName: String, id: Int): Boolean {
        if (isOpened) {
            var bRet = false
            val temp = "SELECT * FROM $strTableName WHERE $idFieldName='$id'"
            val cursor = db.rawQuery(temp, null)
            if (cursor != null) {
                if (cursor.count > 0) {
                    bRet = true
                }
                cursor.close()
            }
            return bRet
        }
        return false
    }

    fun isExistData(strTableName: String, idFieldName: String, id: Int, userfieldName: String, userid: String): Boolean {
        if (isOpened) {
            var bRet = false
            val temp = "SELECT * FROM $strTableName WHERE $idFieldName='$id' AND $userfieldName='$userid'"
            val cursor = db.rawQuery(temp, null)
            if (cursor != null) {
                if (cursor.count > 0) {
                    bRet = true
                }
                cursor.close()
            }
            return bRet
        }
        return false
    }

    fun startTransaction() {
        if (isOpened) {
            db.beginTransaction()
        }
    }

    fun transactionSuccess() {
        if (isOpened) {
            db.setTransactionSuccessful()
        }
    }

    fun endTransaction() {
        if (isOpened) {
            db.endTransaction()
        }
    }

    fun compileStatement(sql: String?): SQLiteStatement {
        return db.compileStatement(sql)
    }
}